hikari/traits.py

```diff
@@ -60,42 +57,64 @@
     from hikari.api import rest as rest_
     from hikari.api import shard as gateway_shard
     from hikari.api import voice as voice_
-    from hikari.events import base_events  # noqa F401 - Unused (False positive)
 
-EventT_co = typing.TypeVar("EventT_co", bound="base_events.Event", covariant=True)
-"""Type-hint for a covariant event implementation instance.
+    _T = typing.TypeVar("_T")
 
-This will bind to the bound event type, or any subclass of that type.
-"""
+_Protocol = NotImplemented
 
-EventT_inv = typing.TypeVar("EventT_inv", bound="base_events.Event")
-"""Type-hint for an invariant event implementation instance.
 
-This will bind to the bound event type only. Subclasses and superclasses will
-not be matched.
-"""
+class _FastProtocol(abc.ABCMeta):
+    @staticmethod
+    def __new__(
+        mcs: typing.Type[_T],
+        cls_name: str,
+        bases: typing.Tuple[typing.Type[typing.Any], ...],
+        namespace: typing.Dict[str, typing.Any],
+    ) -> _T:
+        global _Protocol
 
-PredicateT = typing.Callable[[EventT_co], typing.Union[bool, typing.Coroutine[typing.Any, typing.Any, bool]]]
-"""Type-hint for an event waiter predicate.
+        if _Protocol is NotImplemented:
+            if cls_name != "Protocol":
+                raise TypeError("First instance of _FastProtocol must be Protocol")
 
-This should be a function or coroutine function that consumes a covariant
-instance of the bound event type and returns a boolean that matches
-`builtins.True` if the event is a match for the waiter, or `builtins.False`
-otherwise.
-"""
+            namespace["_is_protocol"] = True
+            namespace["_attributes_"] = ()
+            return (_Protocol := super().__new__(mcs, cls_name, bases, namespace))
 
-# Fixme: shouldn't this be using covariance instead of invariance?
-AsyncCallbackT = typing.Callable[[EventT_inv], typing.Coroutine[typing.Any, typing.Any, None]]
-"""Type-hint for an asynchronous coroutine function event callback.
+        if _Protocol not in bases:
+            namespace["_is_protocol"] = False
 
-This should consume a single argument: the event that was dispatched.
+        else:
+            attributes = tuple(name for name in namespace if not name.startswith("_"))
```

Comment:
While this approach is definitely a lot more efficient, this does seem to slightly violate how are checked against objects and types (since the reference impl does count private and protected attributes, would it be worth it making this better match the behaviour seen in the typing module by using typing.EXCLUDED_ATTRIBUTES to explicitly blacklist attributes (and also maybe checking against type hints as well)?

Reply:
`typing.Protocol` doesnt check for typehints anyways, so including it might break something downstream. And will change the attribute check :+1:

Reply:
> typing.Protocol doesnt check for typehints anyways

it does check against a class's annotations for instance variables

```python
def _get_protocol_attrs(cls):
    """Collect protocol members from a protocol class objects.

    This includes names actually defined in the class dictionary, as well
    as names that appear in annotations. Special names (above) are skipped.
    """
    attrs = set()
    for base in cls.__mro__[:-1]:  # without object
        if base.__name__ in ('Protocol', 'Generic'):
            continue
        annotations = getattr(base, '__annotations__', {})
        for attr in list(base.__dict__.keys()) + list(annotations.keys()):
            if not attr.startswith('_abc_') and attr not in EXCLUDED_ATTRIBUTES:
                attrs.add(attr)
    return attrs
    ```